### DISCUSSION TITLE
3 axis LSM303

### DISCUSSION THREAD

**Discussion by kallepe74** [jeff-burright/Autopilot_ESP32_wifi#7] [Q&A]
Hi Jeff, and thank you very much for an inspiring project. 

Being 50 years old and not done any actual programming for 30 years or so is a challenge. I am still on the lab board stage but I am about to setup a test rig to see if real world acting of the motor output. 

I do have a question on the LSM303 part. Mine seem to be a 3 axis variant without any gyro abilities. To get any output on the webpage I have commented out the Read_Gyro(); in the main ino line 1294 and AP_Loop.ino line 41. That took a while to realize but it also gave some insights in what is going on in the various parts of the code. 

So, having blocked the gyro reading, do you think of any other parts of the code that i should alter compensating for the none existing gyro inputs? When reading about the LSM303 capabilities it should be fit for purpose for this task but I could of course also replace it with a 9 axis variant 

Thanks again for sharing your work

**Comment by jeff-burright** [comment]
Hello, I recall I had a similar issue when I accidentally bought a 6-axis IMU instead of a 9-axis. I sympathize with all the different locations in the code that reference various parts of the compass and how hard it is to keep track. Without the gyro, I don't think you'll be able to have tilt compensation for your heading using this code, so even if you could get it to work I don't think it would serve well. It's not as simple as commenting out the read.gyro function, because other parts of the code ask for the gyro data to calculate pitch/roll/yaw for tilt compensation (as I understand it having cobbled together other libraries and projects to create this one). One thing you might try is to change the user input LSM_LIB to 1 instead of 0, which would use the standard library tilt compensation function that comes with the LSM303 instead of the pilot's compass code. I have no idea if it'll work, but you could try. The simplest path forward and best use of time would probably be to just get a 9DOF version of the IMU. 

One of my winter projects was going to be to add in support for the MPU9250 IMU, which is a newer chip. The one I ordered was a dud though and doesn't read the magnetometer, so it'll be a while. I know I could stand to clean up the code a little more too. 

Out of curiosity, are you building this for a tiller or wheel? Do you plan to use any buttons or remotes, or just your phone? I've just coded control using a small 433mhz remote control as an experiment, but the range on my module is poor so it's not ready to be added to the repository yet.

**Comment by kallepe74** [comment]
Ok, that was what I suspected. Sound like it is better to order a 9DOF then. I did get output at least at the webpage and also in serial interface (with some  println I added to monitor/troubleshot) that it was giving right or left rudder until heading was within the deadband. But this was on the other hand with the LSM303 laying flat on a chair just being spun around not being tilted or anything. 

I will use this for a tiller. My initial idea has been to use a phone for interfacing with the webpage and if it turns out to be something I tend to use a lot maybe move towards buttons or a remote. I sail alone quite frequently and then perhaps a remote would be good.

Thanks for your feedback!
